### PR TITLE
class IteratorOfMapsDataIterator

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
class IteratorOfMapsDataIterator
DataIteratorUtil.mapTransformer()

#### Related Pull Requests
- 

#### Changes
-

### CODE REVIEW

**Comment by labkey-matthewb** [comment]
Some useful (but not yet used) updates to DataIterator. 
* IteratorOfMapsDataIterator in addition to the existing ListOfMapsDataIterator
* mapTransformer() which allows a dev to tranform streams using Map.get()/Map.put() instead of DataIterator.get(int). (slower but useful for converting existing code).

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java [hunk 1]
```diff
@@ -486,8 +486,8 @@ public List<Map<String, Object>> updateRows(User user, Container container, List
             results = super.updateRows(user, container, rows, oldKeys, errors, configParameters, extraScriptContext);
 
             /* setup mini dataiterator pipeline to process lineage */
-            DataIterator di = _toDataIterator("updateRows.lineage", results);
-            ExpDataIterators.derive(user, container, di, true, _sampleType, true);
+            DataIterator dib = _toDataIteratorBuilder("updateRows.lineage", results).getDataIterator(new DataIteratorContext());
+            ExpDataIterators.derive(user, container, dib.getDataIterator(new DataIteratorContext()), true, _sampleType, true);
```
Comment (labkey-jeckels) [inline experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java]: Do we need both calls to `getDataIterator()` (one on each line)?

File: api/src/org/labkey/api/dataiterator/AbstractMapDataIterator.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        {
+            if (assertsEnabled)
+            {
+                ArrayList<Map<String, Object>> copy = new ArrayList<>(rows.size());
+                for (Map<String, Object> row : rows)
+                {
+                    // assumes all ArrayListMaps are case insensitive
+                    assert row instanceof CaseInsensitiveMapWrapper || row instanceof ArrayListMap : "all rows must be either CaseInsensitiveMapWrapper or ArrayListMap";
+                    if (row instanceof ArrayListMap)
+                        ((ArrayListMap) row).setReadOnly(true);
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/dataiterator/AbstractMapDataIterator.java]:
```suggestion
                    if (row instanceof ArrayListMap listMap)
                        listMap.setReadOnly(true);
```

File: api/src/org/labkey/api/dataiterator/DataIteratorUtil.java [hunk 3]
```diff
@@ -387,6 +393,75 @@ public static MapDataIterator wrapMap(DataIterator in, boolean mutable)
     }
 
 
+    /**
+     * This Function<> mechanism is very simple, but less efficient that an operator that takes an input map, and
+     * a pre-created ArrayListMap for output.  For efficiency there are other DataIterator base classes to use.
+     * Should return ArrayListMap or CaseInsensitiveMap.
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:399]: Is this enforced anywhere? If not, an assert or two would help ensure compliance.

**Review by labkey-jeckels — APPROVED** [review]

File: experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java (hunk 1 above)
Comment (labkey-matthewb) [reply experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java]: Hmm, cut-n-paste snafu???  It's a noop, but that's a mistake.

File: api/src/org/labkey/api/dataiterator/DataIteratorUtil.java (hunk 3 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:399]: Not yet...

**Pull request merged**